### ISSUE TITLE
Feature request: sparse matrix max(axis)

### ISSUE THREAD

**Issue by davidbau** [pytorch/pytorch#4247]
The current sparse support is pretty nice!  However, there are some important of collation operations that can not be efficiently be composed out of matmul or operations on _values().  One that is in the way of my current work is axis-wise max.  Ideally:

```
vec = sparsemat.max(axis=None, zero=0)
```

Where an optional zero argument could be used to specify the interpretation of unstored matrix entries.  (E.g., could pass -Infinity if that is the intent.)  In my application, I am using a sparse matrix to store only the significant log probabilities in a grid, and max() with a -Infinity default is the key operation needed to implement a stable row-wise logsumexp().)

For reference: scipy.sparse's inclusion of max is here, although they miss the need for a zero default.
https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.coo_matrix.max.html#scipy.sparse.csr_matrix.max

**Comment by SsnL** [comment]
I agree that there should be `.max`, but I don't agree with the `zero=` argument. It is breaking the definition of sparse matrices that unstored values are zeros.

**Comment by gasse** [comment]
I am facing a problem where `zero=` argument for sparse matrices would be very useful (though I would rather name it `empty=`). I want to perform a softmax on the rows of a sparse matrix, where empty cells shall correspond to 0 probability (hence -Infinity values before softmax). This is useful when implementing graph attention networks (https://arxiv.org/abs/1710.10903) on large graphs, where adjacencies shall be represented as sparse matrices.

Although I am can not think of a proper way to implement this `empty=` option, I could definitely use a `sum()` and a `max()` operation on sparse matrices as well!

**Comment by gasse** [comment]
By the way, a workaround to do summation over a specific dimensions of a sparse vector is as follows:

```python
x = torch.sparse.FloatTensor(
    indices=torch.LongTensor([[0, 1, 1], [2, 0, 2]]),
    values=torch.FloatTensor([3, 4, 5]),
    size=[2, 3])

# sum over second dimension
x = torch.sparse.FloatTensor(
        indices=torch.stack([
            x._indices()[0],
            torch.LongTensor(1).zero_().expand_as(x._indices()[0]),
        ]),
        values=x._values(),
        size=[x.shape[0], 1])

print(x)
print(x.coalesce())  # here summation is actually done
print(x.to_dense())  # here too
```

**Comment by cpuhrsch** [comment]
We just landed the [masked amax function](https://github.com/pytorch/pytorch/pull/65990). Next steps include support for sparse layouts, but I don't want to set false expectations for a timeline here.

**Comment by smorad** [comment]
@cpuhrsch does this mean we won't see a sparse max implementation for the foreseeable future?

**Comment by cpuhrsch** [comment]
Stevemo we have torch._masked.max. We just need to add support for COO and CSR. For now you can also use [pytorch_sparse reductions](https://github.com/rusty1s/pytorch_sparse/blob/master/torch_sparse/reduce.py) which have masking semantics and support for COO, CSR and CSC.
